### ISSUE TITLE
[Question] Can elevators have a roof?

### ISSUE THREAD

**Issue by ExDomino** [McJty/RFTools#1308]
Hello, my question is simple : can elevators have a roof? If no, can you add this as a suggestion? I suggest that the roof would be either at 3 or 5 of elevation from the floor.

**Comment by ExDomino** [comment]
I'm still asking about this important question. I really would like to have elevator with roof.
You could also add wall, and maybe not elevator doors that can be added in the cabin and automatly open when the elevator reach its level? And re-close a short bit after, like true elevator doors. Open on each sides like with MalisisDoor. And in the cabin, in the elevator controler, have the way to open/close the elevator doors at the level cabin. Why not? I know it's maybe hard to program but like that, elevators would be really realistic. ^^

**Comment by McJty** [comment]
It is not that easy however. The entire code is based on this a single platform. In the future I can consider extending this but not in the near future

**Comment by ExDomino** [comment]
Thanks. To help you to code that, you can consider that the roof is always at the same level from the floor : 3 meters, even for the 5x5 elevator. If another 3x3 platform is found 3 meters above the main platform, it moves as well, but cannot reach a floor that cannot handle the roof.

**Comment by winsrp** [comment]
well... I saw that it already takes into consideration only blocks of a single type, how about taking all connected blocks of a single type and you can make the elevator any form you want?

**Comment by McJty** [comment]
That's not the problem. The problem is that my 'smooth elevator movement' technology is only written to hangle a single rectangular platform of blocks

**Comment by winsrp** [comment]
well, ain't those fun little challenges what makes programming fun =D

**Comment by DeFazer** [comment]
Oh! You may want to take a look at this mod: https://minecraft.curseforge.com/projects/thuts-elevators
It features custom-built elevators with walls and roofs, and it looks so-o smooth. Best of all, it's open source, so you could see how exactly that was made so it'd be much easier reimplementing it if you decided to go with this suggestion. Or even borrow some code, if the license allows it. I wanted to use this for my elevator-based needs, but... meh. Aside from rendering, it didn't turn out to be that good.

**Comment by DeFazer** [comment]
Also, I'd like to opt against automatically opening, special doors for elevators. This feels... wrong. So not Minecraft-ish. If you want automatic doors, you set up a redstone system, put up a few transmitters here and there, maybe add a timer and a sensor, and so on. If you want them to smoothly slide apart, you install MalisisDoors or make a really good piston system. I believe that making such things by yourself is an integral part of engineering in Minecraft.

**Comment by ExDomino** [comment]
No ads please, thuts-elevators is too bugged.

I suggest to make that if there is a roof of the same type of the floor, 3 blocks above the floor (4 blocks for 5x5 elevators), it move with the floor. When I say 3 blocks above, I mean that if the elevator was a room, it has 3 blocks of height, or 4 for 5x5 elevators. That would be perfect. No need for walls, just a ceilling. ^_^

**Comment by DeFazer** [comment]
Oh! Sorry if it looked like I'm trying to propose using that mod instead; I actually addressed that to McJty. Since he stated the problem with creating complex elevators is in the rendering code, I suggested looking at that mod's repository and see how it's done there. Because you have to admit - despite its downsides, the rendering is done exceptionally good. You could make an elevator from custom blocks, different materials, decorative fences... Doing just the roof "of the same material 3 blocks above", while nice too, seems quite... incomplete and too specific to me. Not to mention problems of backwards compatibility - imagine someone has his/her/its top floor exactly 3 blocks below the ceiling of the same material as the elevator. If this suggestion was implemented, the next time they update the mod, the elevator would chomp away a piece of their ceiling. 
Don't get me wrong; this is actually a good suggestion! I just thought that it would be better not to do this half-way and make elevators actually customizable, with roofs, walls, doors and everything. On the other hand, that indeed would potentially be a lot more work compared to just creating a second rectangular platform, even with the sample rendering code from that Thut' mod. If McJty decides to implement this, but stick with roof platforms for now (and that's completely understandable), I'd like to at least suggest making the height customizable and the whole roof thing toggleable in the elevator's GUI. This would avoid the backwards compatibility problems, while still providing a nice new feature. ^^

**Comment by Dimentive** [comment]
Do you have an update on this, @McJty? Elevators with roofs are my dream!

**Comment by McJty** [comment]
I'm afraid this is currently very low priority as I have no time and TONS of other things to do. Like I'm in the middle of porting all my mods to 1.14

**Comment by Dimentive** [comment]
Of course. That’s understandable. Hopefully, this is a feature to consider for the future, though, when all the porting is done.
